### DISCUSSION TITLE
Does the Guardian network have a BFT mechanism?

### DISCUSSION THREAD

**Discussion by junha1** [wormhole-foundation/wormhole#2537] [Developer Q&A]
I understood that guardians communicate each other using peer-to-peer networking,
but do they have a finalization mechanism that linearly orders VAAs?

**Comment by kcsongor** [comment]
Let me try to unpack the question here. It sounds like you're asking about whether the guardian network's BFT property is _synchronous_? Guardians produce and broadcast their signatures fully asynchronously, which means that there's no guarantee that a message that was sent in an earlier block will have its signed VAA available earlier.

VAAs are partially ordered based on their origin chain & timestamp. Since there's no global guarantee about any pairwise blockchain timestamp synchronisation, VAA timestamp ordering is only defined between VAAs that original from the same chain.

There is an additional caveat, which is the source chain's finality. Wormhole supports sending messages before the chain finalises (chain-specific details can be found here https://book.wormhole.com/wormhole/3_coreLayerContracts.html?highlight=consistency#consistency-levels). This means that it's possible to observe messages from two different forks of a chain, which again naturally don't have an ordering. Applications that rely on finality (such as the portal token bridge) always require the sending block to finalise before the VAA is emitted, so their VAAs do have a total order from a given chain.

**Reply by junha1** [reply]
Then, the guardian network does not guarantee that two messages from a same chain will be executed in the counterparty chain sequentially in the same order that the two messages are finalized in the source chain? (because it rather depends on the broadcast latency and relayer's behavior?) I'm not sure whether it really matters, but I think it is feasible to guarantee such ordering if the guardians have a consensus algorithm to reach an ordering of VAAs.

**Reply by kcsongor** [reply]
The guardian network does not guarantee anything about message delivery. There is a sequence number field in each VAA that is an incremental identifier of the messages originating from a given emitter which the receiver can check if it's important that they're delivered in order.

**Reply by junha1** [reply]
Ok, got it. Thanks.
